### ISSUE TITLE
Reconsider using url-encoded elements

### ISSUE THREAD

**Issue by Glamdring** [sakurity/securelogin-spec#2]
I'm implementing Java server verification, and I already hit two issues with the chosen token format:

- inevitably people will try to pass the token as a GET parameter instead of a POST request body, which screws the commas - at least when I tried jQuery, on the server (Spring MVC) I got all the commas automatically decoded. Depending on the client and server-side frameworks, these automatic decoding can happen for GET params but sometimes also for POST

- Ruby URI.decode(..) does not replace + with whitespace (https://github.com/sinatra/sinatra/issues/463), but Java URLDecoder does. And since + is part of the Base64 alphabet, this will be causing confusion (I can, of course, pre-replace + with its URL-encoded form before going through the decoding phase, but that's hacky)

Since URI-encoding is meant for URIs, I think it should be reconsidered as a way to encode arbitrary elements such as multi-element tokens. Differences in implementations and "clever" web frameworks will add confusion and issues.

An easy change is to simply replace the commas with pipes (|), and the issues will go away. In fact, apart from making the implementation friendlier in Ruby, I think grouping the signatures and keys is not necessary, so just having a sequence of all the elements, comma-separated should be fine (except the message, which needs to be available in whole for the signature verification)

**Comment by homakov** [comment]
Good catch, yes the url encoding idea isn't best part of the protocol. Maybe it's worth using a JSON array stringify, not sure about | instead of a comma.

Also, I have decided to remove the hmac part from keys and signatures. It's confusing and useless at the moment (could be useful at some point later as shared secret).

So it will be flat as w/o nested csv arrays and I will do something about broken URL encoding.

**Comment by Glamdring** [comment]
Cool. Unrelated: Maybe also add the email to the signed message?

**Comment by homakov** [comment]
Email is a complimentary part, like an account label. By signing it you don't prove that you own it.

**Comment by Glamdring** [comment]
I just realized the change password endpoint uses a GET parameter, so currently I can't have a proper spring-mvc/java implementation of that part (unless I manually re-encode the particular commas).
